### PR TITLE
Add logging to all cache sync waits

### PR DESCRIPTION
**Pull request by howardjohn**
This is to aide in debugging. At debug level, we get high verbosity about all caches. At info level, we get logs about slow syncs and failed syncs.

**Please provide a description of this PR:**

### CODE REVIEW

**Review by zirain — APPROVED** [review]

**Review by hanxiaop — APPROVED** [review]

**Review by ramaraochavali — APPROVED** [review]

File: pkg/kube/client.go [hunk 1]
```diff
@@ -669,12 +670,21 @@ func WaitForCacheSync(stop <-chan struct{}, cacheSyncs ...cache.InformerSynced)
...
 		return true
 	}
+	attempt := 0
+	defer func() {
+		if r {
+			log.WithLabels("name", name, "attempt", attempt, "time", time.Since(t0)).Debugf("sync complete")
+		} else {
+			log.WithLabels("name", name, "attempt", attempt, "time", time.Since(t0)).Errorf("sync failed")
+		}
+	}()
```
Comment (hzxuzhonghu) [inline pkg/kube/client.go:680]: I think this may cause duplication logs printed

File: pilot/pkg/bootstrap/server.go [hunk 2]
```diff
@@ -809,7 +809,7 @@ func (s *Server) addTerminatingStartFunc(name string, fn server.Component) {
 func (s *Server) waitForCacheSync(stop <-chan struct{}) bool {
 	start := time.Now()
 	log.Info("Waiting for caches to be synced")
-	if !kubelib.WaitForCacheSync(stop, s.cachesSynced) {
+	if !kubelib.WaitForCacheSync("server", stop, s.cachesSynced) {
```
Comment (hzxuzhonghu) [inline pilot/pkg/bootstrap/server.go:812]: more explicit: bootstrap server

**Review by jacob-delgado — APPROVED** [review]

**Review by stevenctl — APPROVED** [review]

**Review by ericvn — APPROVED** [review]

File: pkg/kube/client.go [hunk 3]
```diff
@@ -669,12 +670,21 @@ func WaitForCacheSync(stop <-chan struct{}, cacheSyncs ...cache.InformerSynced)
 		}
 		return true
 	}
+	attempt := 0
+	defer func() {
+		if r {
+			log.WithLabels("name", name, "attempt", attempt, "time", time.Since(t0)).Debugf("sync complete")
```
Comment (costinm) [inline pkg/kube/client.go:676]: Not sure if you remove Debug() from the API - but if using the structured mode I don't think Printf should be used. Not a huge deal if we move later to upstream slog.

Comment (costinm) [reply pkg/kube/client.go:676]: Also - we may want to discuss defining some constants for the labels so we have more consistency ( not only for the structured logs, but also for metrics and istiod-generated traces when we add them).

**Review by shankgan — APPROVED** [review]

**Pull request merged**